godot-bindings/src/godot_json.rs

```diff
@@ -0,0 +1,154 @@
+            major: self.version_major,
+            minor: self.version_minor,
+            patch: self.version_patch,
+            status: self.version_status,
+            custom_rev: None,
+        }
+    }
+}
+
+pub fn load_custom_gdextension_json() -> String {
+    let path = std::env::var("GODOT4_GDEXTENSION_JSON")
+        .expect("gdext with `api-custom-json` feature requires GODOT4_GDEXTENSION_JSON environment variable (with the path to the said json).");
+    let json_path = Path::new(&path);
+
+    fs::read_to_string(json_path).unwrap_or_else(|_| {
+        panic!(
+            "failed to open file with custom GDExtension JSON {}.",
+            json_path.display()
+        )
+    })
+}
+
+fn get_gdextension_json_file() -> File {
+    let path = std::env::var("GODOT4_GDEXTENSION_JSON")
+        .expect("gdext with `api-custom-json` feature requires GODOT4_GDEXTENSION_JSON environment variable (with the path to the said json).");
+    let json_path = Path::new(&path);
+
+    File::open(json_path).unwrap_or_else(|_| {
+        panic!(
+            "failed to open file with custom GDExtension JSON {}.",
+            json_path.display()
+        )
+    })
+}
+
+// GDExtension JSON can get excessively big, so we extract only the very first key – the header.
+fn extract_gdextension_json_header_section(
+    mut lines: impl Iterator<Item = Result<String, std::io::Error>>,
+) -> String {
+    let mut header = String::new();
+    'outer: loop {
+        let Some(Ok(line)) = lines.next() else {
+            panic!("Failed to extract \"header\" section from provided GDExtension JSON.")
+        };
+
+        // Keep parsing until we stumble on "header" key.
+        if !line.contains("\"header\"") {
+            continue;
+        }
+        header.push('{');
+
+        // Parse header section, assuming its structure is the same as one in `GDExtensionJSONHeader`.
+        '_parse_header: loop {
+            let Some(Ok(line)) = lines.next() else {
+                panic!("Failed to extract \"header\" section from provided GDExtension JSON.");
+            };
+            if line.contains("}") {
+                header.push('}');
+                break 'outer;
+            }
+            header.push_str(&line);
+        }
+    }
+
+    header
+}
+
+pub(crate) fn read_godot_version() -> GodotVersion {
+    let lines = BufReader::new(get_gdextension_json_file()).lines();
+    let header = extract_gdextension_json_header_section(lines);
+
+    let result: GDExtensionJSONHeader = DeJson::deserialize_json(&header)
+        .unwrap_or_else(|_| panic!("failed to deserialize JSON {}.", header));
+    result.into_godot_version()
+}
+
+pub(crate) fn write_gdextension_headers(out_h_path: &Path, out_rs_path: &Path) {
+    let path = std::env::var("GODOT4_GDEXTENSION_HEADERS")
+        .expect("gdext with `api-custom-json` feature requires GODOT4_GDEXTENSION_HEADERS environment variable (with the path to the said json).");
```

Comment:
I'll fix it later while squashing etc:

```suggestion
        .expect("gdext with `api-custom-json` feature requires GODOT4_GDEXTENSION_HEADERS environment variable (with the path to the said C header).");
```